components/places/android/src/main/java/org/mozilla/places/PlacesConnection.kt

```diff
@@ -17,40 +17,89 @@ import java.util.concurrent.atomic.AtomicLong
 import java.util.concurrent.atomic.AtomicReference
 
 /**
- * An implementation of a [PlacesAPI] backed by a Rust Places library.
+ * An implementation of a [PlacesApi] backed by a Rust Places library.
  *
  * @param path an absolute path to a file that will be used for the internal database.
  * @param encryption_key an optional key used for encrypting/decrypting data stored in the internal
  *  database. If omitted, data will be stored in plaintext.
  */
-class PlacesConnection(path: String, encryption_key: String? = null) : PlacesAPI, AutoCloseable {
+class PlacesApi(path: String, encryption_key: String? = null) : PlacesApiInterface, AutoCloseable {
     private var handle: AtomicLong = AtomicLong(0)
-    private var interruptHandle: InterruptHandle
 
     init {
-        try {
-            handle.set(rustCall { error ->
-                LibPlacesFFI.INSTANCE.places_connection_new(path, encryption_key, error)
-            })
-        } catch (e: InternalPanic) {
-
-            // Places Rust library does not yet support schema migrations; as a very temporary quick
-            // fix to avoid crashes of our upstream consumers, let's delete the database file
-            // entirely and try again.
-            // FIXME https://github.com/mozilla/application-services/issues/438
-            if (e.message != "sorry, no upgrades yet - delete your db!") {
-                throw e
+        handle.set(rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_api_new(path, encryption_key, error)
+        })
+    }
+
+    companion object {
+        // These numbers come from `places::db::ConnectionType`
+        private const val READ_ONLY: Int = 1
+        private const val READ_WRITE: Int = 2
+    }
+
+    override fun open_reader(): ReadablePlacesConnection {
+        val conn_handle = rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_connection_new(handle.get(), READ_ONLY, error)
+        }
+        return ReadablePlacesConnection(conn_handle);
+    }
+
+    override fun open_writer(): WritablePlacesConnection {
```

Comment:
I think it might be worth keeping a writer (if any exist) as a local property, and returning the same reference rather than error on trying to open two.